FWCore/Framework/interface/ESProducer.h

```diff
@@ -152,19 +157,23 @@ namespace edm {
                          TReturn (T ::* iMethod)(const TRecord&),
                          const TArg& iDec,
                          const es::Label& iLabel = {}) {
+      const auto id = static_cast<unsigned int>(numberOfFactories());
```

Comment:
Possibly I am just not understanding something here. What does the numberOfFactories have to do with the transitionID getting passed into the Callback?

Reply:
Each factory generates a proxy for one and only one transition (i.e. EventSetup record). So the number of factories is just a way to identify which proxy is being called.

Reply:
What is worrying me is id is being passed into the constructor for the Callback immediately below this point (you have to scroll far to the right to see it in my browser). That is supposed to be a transition ID in the Callback constructor, not a Proxy identifier. Where id is used below that is probably good, but where it is passed into the Callback just seems incorrect. What if there are 10 DataProxy's? Then the value would be outside the range of the enum for transition IDs.

Reply:
One alternative would be to propagate the transition ID as an argument through all the various get* function calls in the stack that leads all the way back to the EventSetup get function call.

Reply:
@wddgit You are thinking that 'transitionID' here is of type 'edm::TransitionID'. It is not. For EDProducers the numeric value I use to keep track of 'transitionID' is the one from 'edm::TransitionID' since EDProducers have 'callbacks' (i.e. virtual functions) for only those transitions. For ESProducers, a 'transitionID' is an identifier for which 'setWhatProduced' is associated with which Callback. I.e. the value of 'transitionID' here is just local to that particular ESProducer. It is say which 'ESConsumesCollector' gathered the information for that given Callback.

Reply:
Now I understand. Thanks.